### PR TITLE
[Kernel] Handle `KernelEngineException` when reading the `_last_checkpoint` file

### PR DESCRIPTION
**Pull request by vkorukanti**
There is an issue with the `CloseableIterator` interface that Kernel is using. Currently, it extends Java's `iterator`, which doesn't throw any exceptions. We use `CloseableIterator` when returning data read from a file or any incremental data access.  Any IOException in `hasNext` or `next` is wrapped in a `UncheckedIOException` or `RuntimeException`. Users of the `CloseableIterator` need to catch for `UncheckedIOException` or `RuntimeException` explicitly and look at the cause if they are interested in the `IOException`. This is not consistent and causes problems for the code that want to handle exceptions like `FileNotFoundException` (subclass of `IOException`) and take further actions.

* Change the `CloseableIterator.{next, hasNext}` contract to expect `KernelEngineException` for any exceptions that occur while executing in the `Engine`.
* Update the `DefaultParquetHandler` and `DefaultJsonHandler` to throw `KernelEngineException` instead of `UncheckedIOException` or `RuntimeException`.
* In the checkpoint metadata loading method, catch `KernelEngineException` and see if the cause is `FileNotFoundException.` If yes, don't retry loading.

### CODE REVIEW

File: kernel/kernel-api/src/main/java/io/delta/kernel/exceptions/KernelEngineException.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ */
+package io.delta.kernel.exceptions;
+
+import io.delta.kernel.engine.Engine;
+
+/**
+ * Throws when the {@link Engine} encountered an error while executing an operation.
+ */
+public class KernelEngineException extends RuntimeException {
+    private static final String msgT = "Encountered an error from the underlying engine " +
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/exceptions/KernelEngineException.java:24]: what does the capital `T` mean in `msgT`?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/Checkpointer.java [hunk 2]
```diff
@@ -239,22 +239,39 @@ private Optional<CheckpointMetaData> loadMetadataFromFile(Engine engine, int tri
...
-            // we can retry until max tries are exhausted
-            return loadMetadataFromFile(engine, tries + 1);
+        } catch (KernelEngineException ex) {
+            Throwable cause = ex.getCause();
+            if (cause instanceof FileNotFoundException) {
+                return Optional.empty(); // there is no point retrying
+            } else if (cause instanceof Exception) {
+                String msg = String.format(
+                        "Failed to load checkpoint metadata from file %s. " +
+                                "It must be in the process of being writing. " +
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/Checkpointer.java:265]: `being written`

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/Checkpointer.java [hunk 3]
```diff
@@ -239,22 +239,39 @@ private Optional<CheckpointMetaData> loadMetadataFromFile(Engine engine, int tri
...
+                                "Retrying after 1sec. (current attempt = %s)",
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/Checkpointer.java:266]: current attempt of X (total attempts)

File: kernel/kernel-api/src/main/java/io/delta/kernel/utils/CloseableIterator.java [hunk 4]
```diff
@@ -34,6 +37,39 @@
...
+     * Returns true if the iteration has more elements. (In other words, returns true if next would
+     * return an element rather than throwing an exception.)
+     *
+     * @return true if the iteration has more elements
+     * @throws KernelEngineException For any underlying exception occurs in {@link Engine} while
+     *                               trying to execute the operation. The original exception is (if
+     *                               any) wrapped in this exception as cause. E.g.
+     *                               {@link IOException} thrown while trying to read from a Delta
+     *                               log file. It will be wrapped in this exception as cause.
+     * @throws KernelException       When encountered an operation or state that is invalid or
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/utils/CloseableIterator.java:51]: for `KernelEngineException` you state i'ts for exceptions that occur in the Engine.

Am I correct in assuming that `KernelException` is for exceptions that occur in the Kernel, not in the engine? Can you add that very short clarification here?

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/LogReplayMetricsSuite.scala [hunk 5]
```diff
@@ -294,6 +305,32 @@ class LogReplayMetricsSuite extends QueryTest
...
+
+        // there should be one checkpoint file at version 10
+        loadScanFilesCheckMetrics(
+          tc,
+          Table.forPath(tc, path),
+          expJsonVersionsRead = 14L to 11L by -1L,
+          expParquetVersionsRead = Seq(10),
+          // we read the checkpoint twice: once for the P &M and once for the scan files
+          expParquetReadSetSizes = Seq(1, 1),
+          expLastCheckpointReadCalls = Some(1)
```
Comment (scottsand-db) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/LogReplayMetricsSuite.scala:329]: worth a simple comment here:
- if the file does not exist, we try reading it once, realize it doesn't exist, and don't try reading it again
- if the file does exist, we succeed reading it the first time

**Review by scottsand-db — APPROVED** [review]
LGTM with minor comments

File: kernel/kernel-api/src/main/java/io/delta/kernel/utils/CloseableIterator.java (hunk 4 above)
Comment (vkorukanti) [reply kernel/kernel-api/src/main/java/io/delta/kernel/utils/CloseableIterator.java:51]: That is correct. Clarified.

**Pull request merged**